lib/memo_wise.rb

```diff
@@ -3,6 +3,49 @@
 require "memo_wise/version"
 
 module MemoWise
-  class Error < StandardError; end
-  # Your code goes here...
+  def initialize(*values)
+    @_memo = {}
+    super
+  end
+
+  def self.prepended(target)
+    class << target
+      # Implements memoization for the given method name.
+      #
+      # The method must be a *pure function* of the value attributes, and
+      # therefore referentially transparent. Replaces the method with a new
+      # method of the same name, which memoizes the result of calling the
+      # original method. See module comment above for a usage example.
+      #
+      # @param method_name [Symbol]
+      #   Name of method for which to implement memoization.
+      def memo_wise(method_name)
+        not_memoized_name = "#{method_name}_not_memoized".to_sym
+        alias_method not_memoized_name, method_name
+        private not_memoized_name
+
+        define_method(method_name) do |*args|
+          key = [method_name, args].freeze
+          @_memo.fetch(key) do
+            @_memo[key] = send(not_memoized_name, *args)
+          end
+        end
+      end
+
+      # Undoes memoization for the given method name. If the method was
+      # previously memoized, it will no longer be memoized.
+      #
+      # @param method_name [Symbol]
+      #   Name of method for which to undo memoization.
+      def reset_memo_wise(method_name)
+        if instance_methods.include?(method_name)
+          not_memoized_name = "#{method_name}_not_memoized".to_sym
+
+          public not_memoized_name
+          alias_method method_name, not_memoized_name
+          undef_method not_memoized_name
```

Comment:
Can we have spec coverage and code to make sure what happens if you call `reset_memo_wise` on a name that you did not previously call `memo_wise` on?

Reply:
I was trying to get at this with the context `"when the method being reset is not defined"`. Are you seeing something else I could add there?

Reply:
I think I missed it! Thanks @jemmaissroff

Reply:
Ok @jemmaissroff I'm closing this comment as you are correct it's already covered by a spec, but I added a new comment asking if we can document the exception raising as part of the YARD doc elsewhere